smtp/smtpc.c

```diff
@@ -440,10 +452,21 @@ smtpc_execute(struct smtpc_request *req, double timeout)
 		++env->stat.failed_requests;
 		smtpc_request_delete(req);
 		return -1;
+	case CURLE_SEND_ERROR:
+	case CURLE_RECV_ERROR:
+		curl_easy_getinfo(req->easy, CURLINFO_RESPONSE_CODE, &longval);
+		req->status = (int) longval;
+		if (strcmp(req->error, "response reading failed") == 0) {
+			req->status = -1;
+		}
```

Comment:
1. First of all request->status must contain some SMTP code. But with this value I meaned that there is some situation when this code does not correspond to reality. For example the next point.
2. I have catched some specific case, when the client side cant get bytes from reposnse, in this case we get `CURLE_RECV_ERROR` and the message `response reading failed`. Without changing status code I receive 250. 
I am not sure that it is a correct behavior. So that looks like:
   status: 250
   reason: response reading failed
3. This way I don't see another workaround of this case.

Reply:
Maybe I would prefer to handle this case as some CURL or OS error, so I would forward it further to default case [here](https://github.com/tarantool/smtp/blob/master/smtp/smtpc.c#L443). What do you think about it?

Reply:
Thanks for the explanation!

It looks as a problem in libcurl: there is no reliable way to obtain an SMTP reply code (and be sure that it is actually sent by server in a response, not something curl-specific). It is especially strange to get 2xx reply code for failure at reading.

If you don't mind to send a bug report to curl (you need to create a pure C reproducer for that) and clarify here that you need to deliver SMTP reply code (and not something else) to a caller, it would be nice to link it from the comment in the code.

We had a problem of this kind with HTTP(S): a write error was reported instead of a more specific error. It was fixed in libcurl: https://github.com/curl/curl/issues/4310.

Regarding the `default` switch case: that's strange to have two error reporting formats. I think we should revisit it. Filed https://github.com/tarantool/smtp/issues/57.

Reply:
I'm working on it currently. I will keep you up to date about this.

Reply:
Filed in https://github.com/curl/curl/issues/8356.